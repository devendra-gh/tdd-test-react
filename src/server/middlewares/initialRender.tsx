/* global __DEV__ */
import { Request, Response, NextFunction } from 'express';
import React from 'react';
import ReactDOM from 'react-dom/server';

import { processCmsResult } from 'server/utils/cmsResultProcessing';

import config from 'config';
import Html, { HtmlProps } from 'server/Html';

import chunksManifest from './chunk-manifest.json';

type ChunksManifest = {
  [key: string]: any;
};

/* global APP_NESTED_PATH */
const baseUrl = APP_NESTED_PATH || '';

/**
 * @param {Object} chunk
 * @param {Set} scripts
 * @returns {undefined}
 */
const addChunk = (chunk: string, scripts: Set<string>) => {
  const chunks: ChunksManifest = chunksManifest;
  if (chunks[chunk]) {
    chunks[chunk].forEach((asset: string) => scripts.add(asset));
  } else if (__DEV__) {
    throw new Error(`Chunk with name '${chunk}' cannot be found`);
  }
};

/**
 * @param {Request} req
 * @param {Response} res
 * @param {Function} next
 * @returns {undefined}
 */
async function initialRender(req: Request, res: Response, next: NextFunction) {
  try {
    const css = new Set();
    const scripts: Set<string> = new Set();
    const scriptsHead: Set<string> = new Set();
    const data: HtmlProps = {
      app: '',
      staticUrl: config.staticUrl,
      analyticsConfig: {
        key: config.analyticsAppKey,
        host: config.analyticsHost,
      },
    };
    const cms = req.cms || {};
    const mode = req.query.mode || req.cookies.mode || '';
    res.cookie('mode', mode);

    // preload config from document store
    if (['v3', 'v4', 'v5'].some(v => req.path.indexOf(`/${v}`) === 0)) {
      const [, , prototypeId] = req.path.split('/');
      scriptsHead.add(`${baseUrl}/pub/workbench/bundle/${prototypeId}`);
    }

    data.mode = `${mode}`.split(',');

    addChunk('client', scripts);

    data.scripts = Array.from(scripts);
    data.scriptsHead = Array.from(scriptsHead);
    data.styles = [{ id: 'css', cssText: [...css].join('') }];
    data.app = {
      apiUrl: config.api.clientUrl,
      url: config.api.clientUrl,
      meta: {
        favicon: `${baseUrl}/favicon.ico`,
      },
      demoLogin: config.demoLogin,
      smartpassData: req.session ? req.session.tammUserInfo : null,
      cms: processCmsResult(cms, req.cmsJourneyInfo),
      featureFlags: config.featureFlags,
    };
    const html = ReactDOM.renderToStaticMarkup(<Html {...data} />);

    res.status(200).send(`<!doctype html>${html}`);
  } catch (err) {
    next(err);
  }
}

export default initialRender;
