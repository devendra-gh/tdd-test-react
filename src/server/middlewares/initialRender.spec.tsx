import React from 'react';
import httpMocks from 'node-mocks-http';

import initialRender from 'server/middlewares/initialRender';

jest.mock('server/utils/logger', () => ({
  getService: () => ({
    error: jest.fn(),
  }),
}));
jest.mock('../Html', () => (...props: any) => (
  <pre data-role="props-list">{JSON.stringify(props, null, 2)}</pre>
));
jest.mock('../../config', () => ({
  api: {
    clientUrl: 'config.api.clientUrlExample',
    demoLogin: 'config.demoLoginExample',
  },
  cms: {
    tammUrl: 'tammUrl',
    tammWorkbenchUrl: 'tammWorkbenchUrl',
  },
}));
jest.mock('./chunk-manifest.json', () => ({
  client: ['client_script1.js', 'client_script2.js'],
}));

const user = {
  emiratesId: 'emiratesIdExample',
  firstName: 'firstNameExample',
  lastName: 'lastNameExample',
  email: 'emailExample',
  phone: 'phoneExample',
  avatarUrl: 'avatarUrlExample',
  nationalityEN: 'nationalityENExample',
};

describe('initial render', () => {
  let req: any;
  let res: any;
  let next: any;

  beforeEach(() => {
    req = httpMocks.createRequest();
    res = httpMocks.createResponse();
    next = jest.fn();

    res.send = jest.fn();
  });

  it('renders page (no user)', async () => {
    await initialRender(req, res, next);

    expect(res.send).toHaveBeenCalled();
    expect(res.send.mock.calls[0][0]).toMatchSnapshot('Page with not user');
    expect(next).not.toHaveBeenCalled();
  });

  it('renders page (no user, with mode)', async () => {
    req.query.mode = 'some';

    await initialRender(req, res, next);

    expect(res.send).toHaveBeenCalled();
    expect(res.send.mock.calls[0][0]).toMatchSnapshot(
      'renders page (no user, with mode)',
    );
    expect(next).not.toHaveBeenCalled();
  });

  it('renders page (with user)', async () => {
    req.user = user;

    await initialRender(req, res, next);

    expect(res.send).toHaveBeenCalled();
    expect(res.send.mock.calls[0][0]).toMatchSnapshot('Page with not user');
    expect(next).not.toHaveBeenCalled();
  });

  it('renders page (with user and session)', async () => {
    req.user = user;
    req.session = {
      tammUserInfo: {
        some: 'smart pass info',
      },
    };

    await initialRender(req, res, next);

    expect(res.send).toHaveBeenCalled();
    expect(res.send.mock.calls[0][0]).toMatchSnapshot('Page with not user');
    expect(next).not.toHaveBeenCalled();
  });

  it('calls next() in case of error', async () => {
    res.send = jest.fn(() => {
      throw new Error();
    });

    await initialRender(req, res, next);

    expect(next).toHaveBeenCalled();
  });
});
