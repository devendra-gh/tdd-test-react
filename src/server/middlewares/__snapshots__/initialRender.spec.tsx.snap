// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render renders page (no user): Page with not user 1`] = `
"<!doctype html><pre data-role=\\"props-list\\">[
  {
    &quot;app&quot;: {
      &quot;apiUrl&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;url&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;meta&quot;: {
        &quot;favicon&quot;: &quot;/favicon.ico&quot;
      },
      &quot;smartpassData&quot;: null,
      &quot;cms&quot;: {
        &quot;journeyInfo&quot;: {},
        &quot;tammUrl&quot;: &quot;tammUrl&quot;,
        &quot;tammWorkbenchUrl&quot;: &quot;tammWorkbenchUrl&quot;
      }
    },
    &quot;analyticsConfig&quot;: {},
    &quot;mode&quot;: [
      &quot;&quot;
    ],
    &quot;scripts&quot;: [
      &quot;client_script1.js&quot;,
      &quot;client_script2.js&quot;
    ],
    &quot;scriptsHead&quot;: [],
    &quot;styles&quot;: [
      {
        &quot;id&quot;: &quot;css&quot;,
        &quot;cssText&quot;: &quot;&quot;
      }
    ]
  },
  {},
  {}
]</pre>"
`;

exports[`initial render renders page (no user, with mode): renders page (no user, with mode) 1`] = `
"<!doctype html><pre data-role=\\"props-list\\">[
  {
    &quot;app&quot;: {
      &quot;apiUrl&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;url&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;meta&quot;: {
        &quot;favicon&quot;: &quot;/favicon.ico&quot;
      },
      &quot;smartpassData&quot;: null,
      &quot;cms&quot;: {
        &quot;journeyInfo&quot;: {},
        &quot;tammUrl&quot;: &quot;tammUrl&quot;,
        &quot;tammWorkbenchUrl&quot;: &quot;tammWorkbenchUrl&quot;
      }
    },
    &quot;analyticsConfig&quot;: {},
    &quot;mode&quot;: [
      &quot;some&quot;
    ],
    &quot;scripts&quot;: [
      &quot;client_script1.js&quot;,
      &quot;client_script2.js&quot;
    ],
    &quot;scriptsHead&quot;: [],
    &quot;styles&quot;: [
      {
        &quot;id&quot;: &quot;css&quot;,
        &quot;cssText&quot;: &quot;&quot;
      }
    ]
  },
  {},
  {}
]</pre>"
`;

exports[`initial render renders page (with user and session): Page with not user 1`] = `
"<!doctype html><pre data-role=\\"props-list\\">[
  {
    &quot;app&quot;: {
      &quot;apiUrl&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;url&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;meta&quot;: {
        &quot;favicon&quot;: &quot;/favicon.ico&quot;
      },
      &quot;smartpassData&quot;: {
        &quot;some&quot;: &quot;smart pass info&quot;
      },
      &quot;cms&quot;: {
        &quot;journeyInfo&quot;: {},
        &quot;tammUrl&quot;: &quot;tammUrl&quot;,
        &quot;tammWorkbenchUrl&quot;: &quot;tammWorkbenchUrl&quot;
      }
    },
    &quot;analyticsConfig&quot;: {},
    &quot;mode&quot;: [
      &quot;&quot;
    ],
    &quot;scripts&quot;: [
      &quot;client_script1.js&quot;,
      &quot;client_script2.js&quot;
    ],
    &quot;scriptsHead&quot;: [],
    &quot;styles&quot;: [
      {
        &quot;id&quot;: &quot;css&quot;,
        &quot;cssText&quot;: &quot;&quot;
      }
    ]
  },
  {},
  {}
]</pre>"
`;

exports[`initial render renders page (with user): Page with not user 1`] = `
"<!doctype html><pre data-role=\\"props-list\\">[
  {
    &quot;app&quot;: {
      &quot;apiUrl&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;url&quot;: &quot;config.api.clientUrlExample&quot;,
      &quot;meta&quot;: {
        &quot;favicon&quot;: &quot;/favicon.ico&quot;
      },
      &quot;smartpassData&quot;: null,
      &quot;cms&quot;: {
        &quot;journeyInfo&quot;: {},
        &quot;tammUrl&quot;: &quot;tammUrl&quot;,
        &quot;tammWorkbenchUrl&quot;: &quot;tammWorkbenchUrl&quot;
      }
    },
    &quot;analyticsConfig&quot;: {},
    &quot;mode&quot;: [
      &quot;&quot;
    ],
    &quot;scripts&quot;: [
      &quot;client_script1.js&quot;,
      &quot;client_script2.js&quot;
    ],
    &quot;scriptsHead&quot;: [],
    &quot;styles&quot;: [
      {
        &quot;id&quot;: &quot;css&quot;,
        &quot;cssText&quot;: &quot;&quot;
      }
    ]
  },
  {},
  {}
]</pre>"
`;
